### PR TITLE
Parse replays for in game time upon completion

### PR DESCRIPTION
**Pull request by Askaholic**
Requires this db change: https://github.com/FAForever/db/pull/199

Uses the replay parser to count the number of elapsed ticks in the replay data once the stream has completed. This is actually quite a cheap operation, especially because the replay data is already in memory.

Game times for desynced games may be unreliable, but there isn't much we can do about that

### CODE REVIEW

File: requirements/main.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 aiomysql
 prometheus_client
 everett[yaml]
+https://github.com/FAForever/faf-scfa-replay-parser/archive/0.1.zip
```
Comment (Wesmania) [inline requirements/main.txt]: Can't we just copy the code into this repo? I don't think replay parsing will gain another python dependency.

File: replayserver/bookkeeping/bookkeeper.py [hunk 2]
```diff
@@ -32,5 +36,35 @@ def build(cls, database, config):
...
+            logger.debug(f"Parsing replay {game_id}")
+            ticks = await self.get_ticks(stream)
+            await self._queries.update_game_stats(game_id, ticks)
         except BookkeepingError as e:
             logger.warning(f"Failed to save replay for game {game_id}: {e}")
+
+    async def get_ticks(self, stream):
+        """Parse the replay and extract the total number of ticks"""
+        loop = asyncio.get_event_loop()
+        return await loop.run_in_executor(
```
Comment (Wesmania) [inline replayserver/bookkeeping/bookkeeper.py]: Why do we need this?

File: replayserver/bookkeeping/bookkeeper.py [hunk 3]
```diff
@@ -32,5 +36,35 @@ def build(cls, database, config):
...
+            None, lambda: _get_replay_ticks(stream.data.bytes()))
+
+
+def _get_replay_ticks(data):
```
Comment (Wesmania) [inline replayserver/bookkeeping/bookkeeper.py]: This should not be in Bookkeeper, but in a separate class passed in as a dependency, so we can unit test properly.

File: replayserver/bookkeeping/bookkeeper.py [hunk 4]
```diff
@@ -32,5 +36,35 @@ def build(cls, database, config):
...
+            CommandStates.Advance,         # For the tick counter
+            CommandStates.VerifyChecksum,  # For desync detection
+            CommandStates.EndGame
+        }
+    )
+    parser.parse()
+
+    # Version 0.1 has a typo in the method name
+    if parser.get_desync_tics():
+        # TODO: How do we handle desynced games?
```
Comment (Wesmania) [inline replayserver/bookkeeping/bookkeeper.py]: Save them anyway, they're still of some value.

File: replayserver/bookkeeping/bookkeeper.py [hunk 5]
```diff
@@ -32,5 +36,35 @@ def build(cls, database, config):
...
+        pass
+
+    return int(parser.tick)
```
Comment (Wesmania) [inline replayserver/bookkeeping/bookkeeper.py]: Why is this needed?

File: replayserver/bookkeeping/database.py [hunk 6]
```diff
@@ -157,6 +158,15 @@ def __init__(self, db):
             'num_players': player_count
         }
 
+    async def update_game_stats(self, game_id, replay_ticks):
```
Comment (Wesmania) [inline replayserver/bookkeeping/database.py]: Is there a reason for putting this info in the db instead of the replay header? Is this meant to be a way to communicate with another service?

File: replayserver/bookkeeping/bookkeeper.py (hunk 5 above)
Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: It probably isn't

File: replayserver/bookkeeping/bookkeeper.py (hunk 3 above)
Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: Yea I was wondering if I needed to make a whole class for it

File: requirements/main.txt (hunk 1 above)
Comment (Askaholic) [reply requirements/main.txt]: I don't see what the downside of doing it this way is. What if someone wants to write another web based replay parsing app?

File: replayserver/bookkeeping/bookkeeper.py (hunk 2 above)
Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: Maybe not. I'll run the profile tests

File: replayserver/bookkeeping/database.py (hunk 6 above)
Comment (Askaholic) [reply replayserver/bookkeeping/database.py]: The whole idea of this is so that the actual in game times can be displayed in the vault (see the DB PR that I linked above). This way it can be queried by the API.

File: requirements/main.txt (hunk 1 above)
Comment (Wesmania) [reply requirements/main.txt]: In Python? Besides, having this code here lets us discard any processing we don't need.

File: replayserver/bookkeeping/bookkeeper.py (hunk 2 above)
Comment (Wesmania) [reply replayserver/bookkeeping/bookkeeper.py]: If the intention was to run body parsing in a separate loop, then it won't work, since replay parser is pure Python code and does not release the GIL. If it was a C extension (like zlib mostly is), then it could do that.

**Comment by Askaholic** [comment]
Running the profiler:
- Without replay parsing, `bookkeeper` doesn't show up in the SVG
- With replay parsing including desync detection
```
bookkeeper:34:save_replay
1.48%
(0.00%)
43x
```
- With replay parsing, but no desync detection
```
bookkeeper:34:save_replay
1.21%
(0.00%)
43x
```

I should also mention that this is an older version of the parser (I just used it because it was the only release) and @norraxx has been doing some work on it since then which has apparently sped it up measurably.

File: replayserver/bookkeeping/bookkeeper.py (hunk 2 above)
Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: Yea, it appears to make no difference when running your profile tests

File: requirements/main.txt (hunk 1 above)
Comment (Askaholic) [reply requirements/main.txt]: Flask and Django are quite popular AFAIK. 

It would be easier to push it to PyPi (this could even be automated in Travis), but I'm not sure what the implications of that would be.

File: requirements/main-pinned.txt [hunk 7]
```diff
@@ -84,6 +84,8 @@ pyyaml==5.1.1 \
     --hash=sha256:ceacb9e5f8474dcf45b940578591c7f3d960e82f926c707788a570b51ba59190 \
     --hash=sha256:fe6a88094b64132c4bb3b631412e90032e8cfe9745a58370462240b8cb7553cd \
     # via everett
+https://github.com/FAForever/faf-scfa-replay-parser/archive/0.1.zip \
+    --hash=sha256:11e9a9ab4c09e664d2c937b880d44fff944807cb7a5a3c3bedf599f7cb7a4357
```
Comment (norraxx) [inline requirements/main-pinned.txt]: I think, we can update for newer version.

File: replayserver/bookkeeping/bookkeeper.py [hunk 8]
```diff
@@ -32,5 +37,27 @@ def build(cls, database, config):
...
+        reader,
+        parse_until_desync=True,
+        parse_commands={
+            CommandStates.Advance,         # For the tick counter
+            CommandStates.EndGame
+        }
+    )
+    parser.parse()
+
+    return parser.tick
```
Comment (norraxx) [inline replayserver/bookkeeping/bookkeeper.py]: i think, i can update, for such thing there will be method on `ReplayBody`.

File: replayserver/bookkeeping/bookkeeper.py [hunk 9]
```diff
@@ -32,5 +37,27 @@ def build(cls, database, config):
...
+
+
+def _get_replay_ticks(data):
+    reader = ReplayReader(data)
+    parser = ReplayBody(
+        reader,
+        parse_until_desync=True,
+        parse_commands={
+            CommandStates.Advance,         # For the tick counter
+            CommandStates.EndGame
```
Comment (norraxx) [inline replayserver/bookkeeping/bookkeeper.py]: Why do we need to parse end game?

File: replayserver/bookkeeping/bookkeeper.py [hunk 10]
```diff
@@ -32,5 +37,27 @@ def build(cls, database, config):
...
         except BookkeepingError as e:
             logger.warning(f"Failed to save replay for game {game_id}: {e}")
+
+    async def get_ticks(self, stream):
+        """Parse the replay and extract the total number of ticks"""
+        return _get_replay_ticks(stream.data.bytes())
+
+
+def _get_replay_ticks(data):
+    reader = ReplayReader(data)
```
Comment (norraxx) [inline replayserver/bookkeeping/bookkeeper.py]: `ReplayBody` without header won't work.
There https://github.com/FAForever/faf-scfa-replay-parser/blob/master/tests/unittests/replay.py#L82

Comment (norraxx) [reply replayserver/bookkeeping/bookkeeper.py]: Just use that method and give the data to `parse`.

File: replayserver/bookkeeping/bookkeeper.py [hunk 11]
```diff
@@ -32,5 +37,27 @@ def build(cls, database, config):
...
+    parser = ReplayBody(
+        reader,
+        parse_until_desync=True,
```
Comment (norraxx) [inline replayserver/bookkeeping/bookkeeper.py]: I don't think, that this flag will work, because, parser doesn't know about `CommandStates.VerifyChecksum`.

File: replayserver/bookkeeping/bookkeeper.py (hunk 3 above)
Comment (norraxx) [reply replayserver/bookkeeping/bookkeeper.py]: This method you can remove and call
```
from replay_parser.replay import parse

...

parse(data)
```

File: requirements/main-pinned.txt (hunk 7 above)
Comment (norraxx) [reply requirements/main-pinned.txt]: I'll do it in some days.

File: replayserver/bookkeeping/bookkeeper.py (hunk 11 above)
Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: Right, I had `VerifyChecksum` before, but then I removed it. Forgot to remove this

File: replayserver/bookkeeping/bookkeeper.py (hunk 10 above)
Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: The data here is only the replay body. At this point the server has already parsed the header and split the data into two parts

File: replayserver/bookkeeping/bookkeeper.py [hunk 12]
```diff
@@ -38,26 +33,7 @@ def build(cls, database, config):
             logger.debug(f"Saved replay {game_id}")
             metrics.saved_replays.inc()
             logger.debug(f"Parsing replay {game_id}")
-            ticks = await self.get_ticks(stream)
+            ticks = get_replay_ticks(stream.data.bytes())
```
Comment (Wesmania) [inline replayserver/bookkeeping/bookkeeper.py]: I'd rather this was injected, even if we don't have unit tests for bookkeeper.

Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: Why?

Comment (Wesmania) [reply replayserver/bookkeeping/bookkeeper.py]: Because unit testability is nice.

Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: You don't need dependency injection in order to write unit tests. I already wrote a basic test for this function

Comment (Wesmania) [reply replayserver/bookkeeping/bookkeeper.py]: I can see a test for `get_replay_ticks`, I'm talking about Bookkeeper. The way it is now, we can't test Bookkeeper itself without testing replay parsing at the same time.

Comment (Askaholic) [reply replayserver/bookkeeping/bookkeeper.py]: The test isn't supposed to verify that the replay parser works correctly, it's supposed to verify that `get_replay_ticks` calls the parser correctly. For example, when testing initially I forgot to call `parser.parse()` after creating the parser. This resulted in there being no errors, but the tick count was always 0. This is the sort of thing the test verifies. 

Really what needs to be tested is that the actual data from `stream.data.bytes()` of a real replay is parsed correctly. This isn't going to happen with a unit test anyways, it will require an integration test. Why would you need to test the bookkeeper without replay parsing, if it's never going to actually run like that? What would be the benefit of that test? You would just be mocking all real work and simply be checking that python code executes in order.

Comment (Wesmania) [reply replayserver/bookkeeping/bookkeeper.py]: I want it mostly out of principle - I tried my best to avoid strong dependencies in code. I could've used the same argument for different classes, but then their behaviour became complex enough to warrant testing..

Comment (norraxx) [reply replayserver/bookkeeping/bookkeeper.py]: `import this`

Comment (Wesmania) [reply replayserver/bookkeeping/bookkeeper.py]: Sure, let me quote you one: `Special cases aren't special enough to break the rules.`

Comment (norraxx) [reply replayserver/bookkeeping/bookkeeper.py]: First of all, this is not special case. In fact, you really don't wanna have any of special cases, you want to have code as simple as it's possible.

And dude, you're making "big project" or "small tiny service"? I think, it's small tiny service, nothing else. Why are you overdoing it?

Now you're searching the excuses do to make things to complicated, but forgetting and omitting basics. (https://github.com/FAForever/java-guidelines/wiki/Contribution-Guidelines#you-are-not-your-code)

And how you wrote it now, I've been spend at least 8 hours reading your code, and trying to understand, what it's doing. And still I didn't get it. 6 classes in one file, totally lack of typing, short variables, functions that you have to read 3 mintues to understand, what's doing, you used some other package libraries, than we use on other projects, different type of configs, i can continue and continue... Now you're trying to have some dependency injections? That code is already not maintainable! So now there is only you who understand what's going on on that server, not others.

Like I would like to help here, but I can't work on it, because I don't wanna to read everything from beginning for 8 hours again and try to understand.

Comment (Wesmania) [reply replayserver/bookkeeping/bookkeeper.py]: None of this has anything to do with the PR. If you believe this codebase is unmaintainable or otherwise bad, discuss it on Zulip and see if others share your opinion.

File: README.md [hunk 13]
```diff
@@ -14,7 +14,7 @@ You can find documentation in the doc directory. It's built with sphinx using
 autodoc, so you'll need to install server dependencies to generate it properly -
 see the 'dev environment' section. Docs can be built with:
 
-`sphinx-build -b html ./doc ./build`
```
Comment (Wesmania) [inline README.md]: These changes were pulled into master already, you can rebase.

File: setup.py [hunk 14]
```diff
@@ -1,6 +1,14 @@
-from setuptools import setup, find_packages
+import re
 
-install_reqs = open('requirements/main.txt').read().splitlines()
+from setuptools import find_packages, setup
+
+# Split up PyPi packages and dependency links
```
Comment (Wesmania) [inline setup.py]: This shouldn't be needed any more.

File: replayserver/bookkeeping/database.py (hunk 6 above)
Comment (Wesmania) [inline replayserver/bookkeeping/database.py]: We could add a unit test for this function.

File: README.md (hunk 13 above)
Comment (Askaholic) [reply README.md]: I tried to do that already, but git just says `Current branch parse-replay-duration is up to date.`

I can even see that the commit hashes are the same from this branch to the master branch so it might just be a bug in github that it's still displaying them

Comment (Askaholic) [reply README.md]: Ah, I made another update to the readme and now it's showing correctly
